File: lib/libesp32/berry/src/be_mem.c
/********************************************************************
** Copyright (c) 2018-2020 Guan Wenliang
** This file is part of the Berry default interpreter.
** skiars@qq.com, https://github.com/Skiars/berry
** See Copyright Notice in the LICENSE file or at
** https://github.com/Skiars/berry/blob/master/LICENSE
********************************************************************/
#include "be_mem.h"
#include "be_exec.h"
#include "be_vm.h"
#include "be_gc.h"
#include <stdlib.h>
#include <string.h>

#define GC_ALLOC    (1 << 2) /* GC in alloc */

#ifdef BE_EXPLICIT_MALLOC
  #define malloc                BE_EXPLICIT_MALLOC
#endif

#ifdef BE_EXPLICIT_FREE
  #define free                  BE_EXPLICIT_FREE
#endif

#ifdef BE_EXPLICIT_REALLOC
  #define realloc               BE_EXPLICIT_REALLOC
#endif

static void* malloc_from_pool(bvm *vm, size_t size);
static void free_from_pool(bvm *vm, void* ptr, size_t old_size);

#define POOL16_SIZE     16
#define POOL32_SIZE     32

BERRY_API void* be_os_malloc(size_t size)
{
    return malloc(size);
}

BERRY_API void be_os_free(void *ptr)
{
    free(ptr);
}

BERRY_API void* be_os_realloc(void *ptr, size_t size)
{
    return realloc(ptr, size);
}

BERRY_API void* be_realloc(bvm *vm, void *ptr, size_t old_size, size_t new_size)
{
    void *block = NULL;
    // serial_debug("be_realloc ptr=%p old_size=%i new_size=%i\n", ptr, old_size, new_size);

    bbool gc_occured = bfalse;   /* if allocation failed, retry after forced GC */
    if (old_size == new_size) { /* the block unchanged, this also captures creation of a zero byte object */
        return ptr;
    }
    /* from now on, block == NULL means allocation failure */

    while (1) {
        /* Case 1: new allocation */
#if BE_USE_PERF_COUNTERS
        vm->counter_mem_alloc++;
#endif
        if (!ptr || (old_size == 0)) {
            block = malloc_from_pool(vm, new_size);
        }
    
        /* Case 2: deallocate */
        else if (new_size == 0) {
#if BE_USE_PERF_COUNTERS
            vm->counter_mem_free++;
#endif
            if (ptr == NULL) { return NULL; }   /* safeguard */
#if BE_USE_DEBUG_GC
            memset(ptr, 0xFF, old_size); /* fill the structure with invalid pointers */
#endif
            free_from_pool(vm, ptr, old_size);
            break;    /* early exit */
        }

        /* Case 3: reallocate with a different size */
        else if (new_size && old_size) {        // TODO we already know they are not null TODO
#if BE_USE_PERF_COUNTERS
            vm->counter_mem_realloc++;
#endif
            if (new_size <= POOL32_SIZE || old_size <=POOL32_SIZE) {
                /* complex case with different pools */
                if (new_size <= POOL16_SIZE && old_size <= POOL16_SIZE) {
                    // no change of slot
                    block = ptr;
                    break;
                } else if (new_size > POOL16_SIZE && old_size > POOL16_SIZE && new_size <= POOL32_SIZE && old_size <= POOL32_SIZE) {
                    // no change of slot
                    block = ptr;
                    break;
                } else {
                    /* one of the buffer is out of pool, the other not */
                    block = malloc_from_pool(vm, new_size);
                    if (block) {
                        /* copy memory */
                        size_t min_size = old_size < new_size ? old_size : new_size;
                        memmove(block, ptr, min_size);
                        // serial_debug("memmove from %p to %p size=%i\n", ptr, block, min_size);
                        free_from_pool(vm, ptr, old_size);
                    }
                }
            } else {
                block = realloc(ptr, new_size);
                // serial_debug("realloc from %p to %p size=%i", ptr, block, new_size);
            }
        } /* end of reallocation */

        /* exit allocator, do we need to GC ? */
        if (block) { break; }               /* all good */
        if (gc_occured) {                   /* already tried gc, can't do much anymore */
            be_throw(vm, BE_MALLOC_FAIL);   /* lack of heap space */
        }
        // serial_debug("be_realloc force_gc\n");
        /* force GC now */
        vm->gc.status |= GC_ALLOC;
        be_gc_collect(vm); /* try to allocate again after GC */
        vm->gc.status &= ~GC_ALLOC;
        gc_occured = btrue;     /* don't try again GC */
    }
    vm->gc.usage = vm->gc.usage + new_size - old_size; /* update allocated count */

    // serial_debug("be_realloc ret=%p\n", block);
    return block;
}

BERRY_API void* be_move_to_aligned(bvm *vm, void *ptr, size_t size) {
    (void)vm;
    (void)size;
#if BE_USE_MEM_ALIGNED
    if (size <= POOL32_SIZE) {
        return ptr;     /* if in memory pool, don't move it so be_free() will continue to work */
    }
    void* iram = berry_malloc32(size);
    if (iram) {
        memcpy(iram, ptr, size);   /* new_size is always smaller than initial mem zone */
        free(ptr);      // TODO gc size is now wrong
        return iram;
    }
#endif
    return ptr;
}

/* Special allocator for structures under 32 bytes */
typedef uint8_t mem16[16];      /* memory line of 16 bytes */
typedef uint8_t mem32[32];      /* memory line of 32 bytes */
#define POOL16_SLOTS   31
#define POOL16_BITMAP_FULL  ((1UL<<POOL16_SLOTS)-1)      /* 31 bits set to 1 - 0x7FFFFFFF */
#define POOL32_SLOTS   15
#define POOL32_BITMAP_FULL  ((1UL<<POOL32_SLOTS)-1)      /* 15 bits set to 1 - 0x7FFF*/

typedef struct gc16_t {
    uint32_t bitmap;            /* bitmap or allocated lines, 0 is free */
    struct gc16_t*  next;
    mem16    lines[POOL16_SLOTS];
} gc16_t;

typedef struct gc32_t {
    uint32_t bitmap;            /* bitmap or allocated lines, 0 is free */
    struct gc32_t*  next;
    mem32    lines[POOL32_SLOTS];
} gc32_t;

static void* malloc_from_pool(bvm *vm, size_t size) {
    if (size == 0) return NULL;
    if (size <= POOL16_SIZE) {
        /* allocate in pool 16 */
        gc16_t* pool16 = vm->gc.pool16;
        while (pool16)  {
            /* look for an empty slot */
            if (pool16->bitmap != 0x0000) {
                /* there is a free slot */
#ifdef __GNUC__
                int bit = __builtin_ffs(pool16->bitmap) - 1;
#else
                int bit = ffs(pool16->bitmap) - 1;
#endif
                if (bit >= 0) {
                    /* we found a free slot */
                    // bitClear(pool16->bitmap, bit);
                    pool16->bitmap &= ~(1UL << bit);
                    // serial_debug("malloc_from_pool found slot in pool %p, bit %i, ptr=%p\n", pool16, bit, &pool16->lines[bit]);
                    return &pool16->lines[bit];
                }
            }
            pool16 = pool16->next;
        }
        /* no slot available, we allocate a new pool */
        pool16 = (gc16_t*) malloc(sizeof(gc16_t));
        if (!pool16) { return NULL; } /* out of memory */
        pool16->next = vm->gc.pool16;
        pool16->bitmap = POOL16_BITMAP_FULL - 1;      /* allocate first line */
        vm->gc.pool16 = pool16;          /* insert at head of linked list */
        // serial_debug("malloc_from_pool allocated new pool %p, size=%i p=%p\n", pool16, sizeof(gc16_t), &pool16->lines[0]);
        return &pool16->lines[0];
    }

    if (size <= POOL32_SIZE) {
        /* allocate in pool 32 */
        gc32_t* pool32 = vm->gc.pool32;
        while (pool32)  {
            /* look for an empty slot */
            if (pool32->bitmap != 0x0000) {
                /* there is a free slot */
#ifdef __GNUC__
                int bit = __builtin_ffs(pool32->bitmap) - 1;
#else
                int bit = ffs(pool32->bitmap) - 1;
#endif
                if (bit >= 0) {
                    /* we found a free slot */
                    // bitClear(pool32->bitmap, bit);
                    pool32->bitmap &= ~(1UL << bit);
                    // serial_debug("malloc_from_pool found slot in pool %p, bit %i, ptr=%p\n", pool32, bit, &pool32->lines[bit]);
                    return &pool32->lines[bit];
                }
            }
            pool32 = pool32->next;
        }
        /* no slot available, we allocate a new pool */
        pool32 = (gc32_t*) malloc(sizeof(gc32_t));
        if (!pool32) { return NULL; } /* out of memory */
        pool32->next = vm->gc.pool32;
        pool32->bitmap = POOL32_BITMAP_FULL - 1;      /* allocate first line */
        vm->gc.pool32 = pool32;          /* insert at head of linked list */
        // serial_debug("malloc_from_pool allocated new pool %p, size=%i p=%p\n", pool32, sizeof(gc32_t), &pool32->lines[0]);
        return &pool32->lines[0];
    }

    return malloc(size);    /* default to system malloc */
}

static void free_from_pool(bvm *vm, void* ptr, size_t old_size) {
    if (old_size <= POOL16_SIZE) {
        gc16_t* pool16 = vm->gc.pool16;
        while (pool16) {
            int32_t offset = (uint8_t*)ptr - (uint8_t*) &pool16->lines[0];
            // serial_debug("free_from_pool ptr=%p pool=%p offset=%i\n", ptr,pool16, offset);
            if ((offset >= 0) && (offset < POOL16_SLOTS*16) && ((offset & 0x0F) == 0)) {
                int bit = offset >> 4;
                // serial_debug("free_from_pool ptr=%p fond pool=%p bit=%i\n", ptr, pool16, bit);
                // bitSet(pool16->bitmap, bit);
                pool16->bitmap |= 1UL << bit;
                return;
            }
            pool16 = pool16->next;
        }
    }
    else if (old_size <= POOL32_SIZE) {
        gc32_t* pool32 = vm->gc.pool32;
        while (pool32) {
            int32_t offset = (uint8_t*)ptr - (uint8_t*) &pool32->lines[0];
            // serial_debug("free_from_pool pool=%p offset=%i\n", pool32, offset);
            if ((offset >= 0) && (offset < POOL16_SLOTS*16) && ((offset & 0x1F) == 0)) {
                int bit = offset >> 5;
                // serial_debug("free_from_pool ptr=%p fond pool=%p bit=%i\n", ptr, pool32, bit);
                // bitSet(pool32->bitmap, bit);
                pool32->bitmap |= 1UL << bit;
                return;
            }
            pool32 = pool32->next;
        }
    }
    else {
        // serial_debug("free_from_pool free=%p\n", ptr);
        free(ptr);
    }
}

BERRY_API void be_gc_memory_pools(bvm *vm) {
    gc16_t** prev16 = &vm->gc.pool16;
    gc16_t* pool16 = vm->gc.pool16;
    while (pool16) {
        if (pool16->bitmap == POOL16_BITMAP_FULL) {
            /* pool is empty, we can free it */
            *prev16 = pool16->next;
            gc16_t* pool_to_freed = pool16;
            pool16 = pool16->next;  /* move to next */
            free(pool_to_freed);
        } else {
            prev16 = &pool16->next;
            pool16 = pool16->next;  /* move to next */
        }
    }
    gc32_t** prev32 = &vm->gc.pool32;
    gc32_t* pool32 = vm->gc.pool32;
    while (pool32) {
        if (pool32->bitmap == POOL32_BITMAP_FULL) {
            /* pool is empty, we can free it */
            *prev32 = pool32->next;
            gc32_t* pool_to_freed = pool32;
            pool32 = pool32->next;  /* move to next */
            free(pool_to_freed);
        } else {
            prev32 = &pool32->next;
            pool32 = pool32->next;  /* move to next */
        }
    }
}

BERRY_API void be_gc_init_memory_pools(bvm *vm) {
    vm->gc.pool16 = NULL;
    vm->gc.pool32 = NULL;
}

BERRY_API void be_gc_free_memory_pools(bvm *vm) {
    gc16_t* pool16 = vm->gc.pool16;
    while (pool16) {
        gc16_t* pool_to_freed = pool16;
        pool16 = pool16->next;
        be_os_free(pool_to_freed);
    }
    vm->gc.pool16 = NULL;

    gc32_t* pool32 = vm->gc.pool32;
    while (pool32) {
        gc32_t* pool_to_freed = pool32;
        pool32 = pool32->next;
        be_os_free(pool_to_freed);
    }
    vm->gc.pool32 = NULL;
}

/* https://github.com/hcs0/Hackers-Delight/blob/master/pop.c.txt - count number of 1-bits */
static int pop0(uint32_t n) __attribute__((unused));
static int pop0(uint32_t n) {
    n = (n & 0x55555555u) + ((n >> 1) & 0x55555555u);
    n = (n & 0x33333333u) + ((n >> 2) & 0x33333333u);
    n = (n & 0x0f0f0f0fu) + ((n >> 4) & 0x0f0f0f0fu);
    n = (n & 0x00ff00ffu) + ((n >> 8) & 0x00ff00ffu);
    n = (n & 0x0000ffffu) + ((n >>16) & 0x0000ffffu);
    return n;
}

#ifdef __GNUC__
  #define count_bits_1(v)   __builtin_popcount(v)
#else
  #define count_bits_1(v)   pop0(v)
#endif


BERRY_API void be_gc_memory_pools_info(bvm *vm, size_t* slots_used, size_t* slots_allocated) {
    size_t used = 0;
    size_t allocated = 0;

    gc16_t* pool16 = vm->gc.pool16;
    while (pool16) {
        allocated += POOL16_SLOTS;
        used += POOL16_SLOTS - count_bits_1(pool16->bitmap);
        pool16 = pool16->next;
    }

    gc32_t* pool32 = vm->gc.pool32;
    while (pool32) {
        allocated += POOL32_SLOTS;
        used += POOL32_SLOTS - count_bits_1(pool32->bitmap);
        pool32 = pool32->next;
    }
    if (slots_used) { *slots_used = used; }
    if (slots_allocated) { *slots_allocated = allocated; }
}

File: lib/libesp32/berry/berry_port/be_mem.py
"""
Berry memory management wrappers.
Mirrors: src/be_mem.c / src/be_mem.h

In the C implementation, be_realloc is the central allocator that tracks
gc.usage and triggers GC when allocation fails. The Python port wraps
Python's native allocation, maintaining the gc.usage counter for GC
threshold logic. Memory pools (pool16/pool32) are skipped — Python's
allocator handles small objects efficiently.
"""

# /********************************************************************
# ** Copyright (c) 2018-2020 Guan Wenliang
# ** This file is part of the Berry default interpreter.
# ** skiars@qq.com, https://github.com/Skiars/berry
# ** See Copyright Notice in the LICENSE file or at
# ** https://github.com/Skiars/berry/blob/master/LICENSE
# ********************************************************************/

from berry_port.be_object import BE_MALLOC_FAIL
from berry_port.berry_conf import BE_USE_DEBUG_GC

# /* GC status flag for "allocation in progress" */
# #define GC_ALLOC    (1 << 2)
GC_ALLOC = (1 << 2)


# ============================================================================
# OS-level allocation wrappers (be_mem.h / be_mem.c)
# ============================================================================

# BERRY_API void* be_os_malloc(size_t size)
# {
#     return malloc(size);
# }
def be_os_malloc(size):
    """Allocate a new block of *size* bytes.

    In C this calls malloc(). In Python we return a bytearray as a
    mutable memory block, or None for zero-size requests.
    """
    if size <= 0:
        return None
    return bytearray(size)


# BERRY_API void be_os_free(void *ptr)
# {
#     free(ptr);
# }
def be_os_free(ptr):
    """Free a previously allocated block.

    In Python, releasing the reference is sufficient — the Python GC
    will reclaim the memory.
    """
    pass  # Python GC handles deallocation


# BERRY_API void* be_os_realloc(void *ptr, size_t size)
# {
#     return realloc(ptr, size);
# }
def be_os_realloc(ptr, size):
    """Reallocate *ptr* to *size* bytes.

    Returns a new bytearray with the content of *ptr* (truncated or
    zero-extended to *size*).
    """
    if ptr is None:
        return be_os_malloc(size)
    if size <= 0:
        return None
    old_len = len(ptr) if ptr is not None else 0
    new_buf = bytearray(size)
    copy_len = min(old_len, size)
    new_buf[:copy_len] = ptr[:copy_len]
    return new_buf


# ============================================================================
# Berry-level allocator with GC tracking (be_mem.c)
# ============================================================================

# BERRY_API void* be_realloc(bvm *vm, void *ptr, size_t old_size, size_t new_size)
# {
#     void *block = NULL;
#     bbool gc_occured = bfalse;
#     if (old_size == new_size) {
#         return ptr;
#     }
#     while (1) {
#         if (!ptr || (old_size == 0)) {
#             block = malloc_from_pool(vm, new_size);
#         }
#         else if (new_size == 0) {
#             if (ptr == NULL) { return NULL; }
#             if (BE_USE_DEBUG_GC || comp_is_gc_debug(vm)) {
#                 memset(ptr, 0xFF, old_size);
#             }
#             free_from_pool(vm, ptr, old_size);
#             break;
#         }
#         else if (new_size && old_size) {
#             block = realloc(ptr, new_size);
#         }
#         if (block) { break; }
#         if (gc_occured) {
#             be_throw(vm, BE_MALLOC_FAIL);
#         }
#         vm->gc.status |= GC_ALLOC;
#         be_gc_collect(vm);
#         vm->gc.status &= ~GC_ALLOC;
#         gc_occured = btrue;
#     }
#     vm->gc.usage = vm->gc.usage + new_size - old_size;
#     return block;
# }
def be_realloc(vm, ptr, old_size, new_size):
    """Central Berry allocator with gc.usage tracking.

    Wraps Python allocation. On failure, triggers a GC cycle and
    retries once before raising BE_MALLOC_FAIL via be_throw.

    Memory pools are skipped — Python's allocator handles small objects.
    """
    if old_size == new_size:
        return ptr

    block = None
    gc_occurred = False

    while True:
        # Case 1: new allocation (ptr is None or old_size == 0)
        if ptr is None or old_size == 0:
            if new_size > 0:
                try:
                    block = bytearray(new_size)
                except MemoryError:
                    block = None
            else:
                block = None

        # Case 2: deallocation (new_size == 0)
        elif new_size == 0:
            # In Python, just drop the reference; nothing to explicitly free.
            # In debug mode the C code memsets to 0xFF — skip in Python.
            block = None
            break  # early exit

        # Case 3: reallocation (both old_size and new_size > 0)
        else:
            try:
                new_buf = bytearray(new_size)
                copy_len = min(old_size, new_size)
                if ptr is not None:
                    new_buf[:copy_len] = ptr[:copy_len]
                block = new_buf
            except MemoryError:
                block = None

        # Check result
        if block is not None:
            break

        if gc_occurred:
            # Already tried GC, can't recover — raise allocation failure.
            # Import here to avoid circular imports at module level.
            from berry_port.be_exec import be_throw
            be_throw(vm, BE_MALLOC_FAIL)

        # Force a GC cycle and retry
        vm.gc.status |= GC_ALLOC
        from berry_port.be_gc import be_gc_collect
        be_gc_collect(vm)
        vm.gc.status &= ~GC_ALLOC
        gc_occurred = True

    # Update allocated-bytes counter
    vm.gc.usage = vm.gc.usage + new_size - old_size
    return block


# ============================================================================
# Convenience macros (be_mem.h)
# ============================================================================

# #define be_malloc(vm, size)         be_realloc((vm), NULL, 0, (size))
def be_malloc(vm, size):
    """Allocate *size* bytes (convenience wrapper around be_realloc)."""
    return be_realloc(vm, None, 0, size)


# #define be_free(vm, ptr, size)      be_realloc((vm), (ptr), (size), 0)
def be_free(vm, ptr, size):
    """Free a block of *size* bytes (convenience wrapper around be_realloc)."""
    return be_realloc(vm, ptr, size, 0)


# ============================================================================
# Memory pool stubs (skipped per design decision)
# ============================================================================

# Memory pools (pool16/pool32) are not implemented in the Python port.
# Python's allocator handles small objects efficiently, so the pool
# infrastructure from be_mem.c is unnecessary. The following functions
# are provided as no-ops for API compatibility.

# BERRY_API void be_gc_init_memory_pools(bvm *vm) {
#     vm->gc.pool16 = NULL;
#     vm->gc.pool32 = NULL;
# }
def be_gc_init_memory_pools(vm):
    """No-op — memory pools are not used in the Python port."""
    pass


# BERRY_API void be_gc_free_memory_pools(bvm *vm) { ... }
def be_gc_free_memory_pools(vm):
    """No-op — memory pools are not used in the Python port."""
    pass


# BERRY_API void be_gc_memory_pools(bvm *vm) { ... }
def be_gc_memory_pools(vm):
    """No-op — memory pools are not used in the Python port."""
    pass


# BERRY_API void be_gc_memory_pools_info(bvm *vm, size_t* slots_used, size_t* slots_allocated)
def be_gc_memory_pools_info(vm):
    """Return (slots_used, slots_allocated) — always (0, 0) in Python port."""
    return (0, 0)
